src/AutoHashEqualsCached.jl

```diff
@@ -69,135 +77,207 @@ function unpack_name(node::Expr)
         where_list = n.args[2:length(n.args)]
         type_params = map(unpack_name, where_list)
         full_type_name = Expr(:curly, type_name, type_params...)
         return (type_name, full_type_name, where_list)
     elseif n.head === :(<:)
-        return unpack_type_name(n.args[1])
+        return unpack_type_name(__source__, n.args[1])
     else
-        error("macro @auto_hash_equals_cached applied to type with unexpected signature: $n")
+        error("$(__source__.file):$(__source__.line): macro @auto_hash_equals_cached applied to type with unexpected signature: $n")
     end
 end
 
-function get_fields(struct_decl::Expr; prevent_inner_constructors=false)
+function get_fields(__source__, struct_decl::Expr; prevent_inner_constructors=false)
     member_names = Vector{Symbol}()
     member_decls = Vector()
 
-    add_field(b) = nothing
-    function add_field(b::Symbol)
+    add_field(__source__, b) = nothing
+    function add_field(__source__, b::Symbol)
         push!(member_names, b)
         push!(member_decls, b)
     end
-    function add_field(b::Expr)
+    function add_field(__source__, b::Expr)
         if b.head === :block
             add_fields(field)
         elseif b.head === :const
-            add_field(b.args[1])
+            add_field(__source__, b.args[1])
         elseif b.head === :(::) && b.args[1] isa Symbol
             push!(member_names, b.args[1])
             push!(member_decls, b)
         elseif b.head === :macrocall
-            add_field(b.args[3])
+            add_field(__source__, b.args[3])
         elseif b.head === :function || b.head === :(=) && (b.args[1] isa Expr && b.args[1].head in (:call, :where))
             # :function, :equals:call, :equals:where are defining functions - inner constructors
             # we don't want to permit that if it would interfere with us producing them.
             prevent_inner_constructors &&
-                error("macro @auto_hash_equals_cached should not be used on a struct that declares an inner constructor")
+                error("$(__source__.file):$(__source__.line): macro @auto_hash_equals_cached should not be used on a struct that declares an inner constructor")
         end
     end
-    function add_fields(b::Expr)
+    function add_fields(__source__, b::Expr)
         @assert b.head === :block
         for field in b.args
-            add_field(field)
+            if field isa LineNumberNode
+                __source__ = field
+            else
+                add_field(__source__, field)
+            end
         end
     end
 
     @assert (struct_decl.args[3].head === :block)
-    add_fields(struct_decl.args[3])
+    add_fields(__source__, struct_decl.args[3])
     return (member_names, member_decls)
 end
 
-"""
-    @auto_hash_equals_cached struct Foo ... end
+function check_valid_alt_hash_name(__source__, alt_hash_name)
+    alt_hash_name === nothing || alt_hash_name isa Symbol || Base.is_expr(alt_hash_name, :.) ||
```

Comment:
Why do we need to support `Base.is_expr(alt_hash_name, :.)`? 🤔

Reply:
Because when you specify the hash function name, we want to permit it to be a qualified name, like `MuttsDict.mutts_hash` in our case.